src/core/lib/promise/mpsc.cc

```diff
@@ -0,0 +1,419 @@
+  auto active_tokens = active_tokens_.load(std::memory_order_relaxed);
+  while (true) {
+    if ((active_tokens & kActiveTokensMask) > max_queued_) {
+      if (active_tokens & kActiveTokensWakerBit) {
+        return false;
+      }
+      if (active_tokens_waker_.load(std::memory_order_relaxed) == nullptr) {
+        // Safe to directly store since we are the only writer.
+        active_tokens_waker_.store(
+            new Waker(GetContext<Activity>()->MakeOwningWaker()),
+            std::memory_order_release);
+      }
+      if (active_tokens_.compare_exchange_weak(
+              active_tokens, active_tokens | kActiveTokensWakerBit,
+              std::memory_order_relaxed)) {
+        return false;
+      }
+    } else {
+      // Observed active tokens less than max_queued_.
+      // This must continue for the duration of this code path since it is the
+      // only incrementer.
+      return true;
+    }
+  }
+}
+
+void Mpsc::DrainMpscFrom(bool wakeup, Node* node) {
+  GRPC_LATENT_SEE_INNER_SCOPE("Mpsc::DrainMpscFrom");
+  while (node != nullptr) {
+    auto next =
+        node->next_.exchange(Node::kWakerPtr, std::memory_order_acquire);
+    if (next & Node::kWakerPtr) {
+      DCHECK_NE(next, Node::kWakerPtr);
+      Waker* waker = reinterpret_cast<Waker*>(next & ~Node::kWakerPtr);
+      if (wakeup) waker->Wakeup();
+      delete waker;
+      next = 0;
+    }
+    if (node != &stub_) ReleaseTokensAndClose(wakeup, node);
+    node = reinterpret_cast<Node*>(next);
+  }
+}
+
+Poll<Mpsc::Node*> Mpsc::Dequeue() {
+  GRPC_LATENT_SEE_INNER_SCOPE("Mpsc::Dequeue");
+  Node* tail = tail_;
+  uintptr_t next = tail->next_.load(std::memory_order_acquire);
+retry_all:
+  if (tail == &stub_) {
+    if (next == 0) {
+      // List is (ephemerally) empty - create a waker so we get woken up
+      // when the queue is not empty.
+      Waker* waker = new Waker(GetContext<Activity>()->MakeNonOwningWaker());
+      if (!tail->next_.compare_exchange_weak(
+              next, reinterpret_cast<uintptr_t>(waker) | Node::kWakerPtr,
+              std::memory_order_acq_rel)) {
+        delete waker;
+        goto retry_all;
+      }
+      return Pending{};  // pending
+    }
+    if (next & Node::kWakerPtr) {
+      // null next waker => list closed
+      DCHECK_NE(next, Node::kWakerPtr);
+      // List is (ephemerally) empty - but we've already asked to be notified
+      // when non-empty.
+      return Pending{};  // pending
+    }
+    tail = reinterpret_cast<Node*>(next);
+    tail_ = tail;
+    next = tail->next_.load(std::memory_order_acquire);
+  }
+  if (next != 0 && (next & Node::kWakerPtr) == 0) {
+    tail_ = reinterpret_cast<Node*>(next);
+    return tail;
+  }
+  Node* head = head_.load(std::memory_order_acquire);
+  // Only the reader can close, so we should not see that here.
+  DCHECK_NE(head, nullptr);
+  if (tail != head) {
+    auto tail_next = head->next_.load(std::memory_order_acquire);
```

Comment:
should this be tail->next_.load(...) ?

Reply:
no, everything following was busted i think